### PR TITLE
[IOTDB-3511] Show Region tool

### PR DESCRIPTION
**Pull request by RYH61**
add new function: show (data | schema)? regions
图片

### CODE REVIEW

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL~

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/ConfigRequestType.java [hunk 1]
```diff
@@ -61,5 +61,6 @@ public enum ConfigRequestType {
   ApplyConfigNode,
   CreateFunction,
   DropFunction,
-  GetNodePathsPartition;
+  GetNodePathsPartition,
+  GetRegionsInfo;
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/ConfigRequestType.java:65]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/read/GetRegionsInfoReq.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  protected void serializeImpl(DataOutputStream stream) throws IOException {
+    stream.writeInt(getType().ordinal());
+    ReadWriteIOUtils.write(regionType, stream);
+  }
+
+  @Override
+  protected void deserializeImpl(ByteBuffer buffer) throws IOException {
+    regionType = ReadWriteIOUtils.readInt(buffer);
+  }
+}
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/read/GetRegionsInfoReq.java:61]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/read/GetRegionsInfoReq.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.iotdb.confignode.consensus.request.ConfigRequestType;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public class GetRegionsInfoReq extends ConfigRequest {
+
+  private int regionType;
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/read/GetRegionsInfoReq.java:32]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/read/GetRegionsInfoReq.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.apache.iotdb.confignode.consensus.request.ConfigRequest;
+import org.apache.iotdb.confignode.consensus.request.ConfigRequestType;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public class GetRegionsInfoReq extends ConfigRequest {
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/read/GetRegionsInfoReq.java:30]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/consensus/response/RegionsInfoResp.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+package org.apache.iotdb.confignode.consensus.response;
+
+import org.apache.iotdb.common.rpc.thrift.TSStatus;
+import org.apache.iotdb.confignode.rpc.thrift.TRegionInfos;
+import org.apache.iotdb.consensus.common.DataSet;
+
+import java.util.List;
+
+public class RegionsInfoResp implements DataSet {
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/consensus/response/RegionsInfoResp.java:28]: fixed~

File: thrift-confignode/src/main/thrift/confignode.thrift [hunk 6]
```diff
@@ -229,6 +229,27 @@ struct TDropFunctionReq {
...
+  1: optional i32 regionType;
+}
+
+struct TRegionInfos {
+  1: required i32 regionId;
+  2: required i32 regionType;
+  3: required string storageGroup
+  4: required list<i32> dataNodeId;
+  5: required list<string> rpcAddresss;
+  6: required list<i32> rpcPort;
```
[the remark this answers is not in the document]
Comment (RYH61) [reply thrift-confignode/src/main/thrift/confignode.thrift:243]: fixed~

File: thrift-confignode/src/main/thrift/confignode.thrift [hunk 7]
```diff
@@ -229,6 +229,27 @@ struct TDropFunctionReq {
...
+  7: required i32 slots;
```
[the remark this answers is not in the document]
Comment (RYH61) [reply thrift-confignode/src/main/thrift/confignode.thrift:244]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 8]
```diff
@@ -685,6 +691,62 @@ public void processLoadSnapshot(File snapshotDir) throws TException, IOException
...
+    regionResp.setRegionInfosList(tRegionInfosList);
+    regionResp.setStatus(RpcUtils.getStatus(TSStatusCode.SUCCESS_STATUS));
+    return regionResp;
+  }
+
+  private void buildTRegionsInfo(
+      List<TRegionInfos> tRegionInfosList,
+      String storageGroup,
+      TRegionReplicaSet replicaSet,
+      RegionGroup regionInfoMap) {
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:727]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 9]
```diff
@@ -685,6 +691,62 @@ public void processLoadSnapshot(File snapshotDir) throws TException, IOException
...
+                      if (replicaSet.getRegionId().getType().ordinal()
+                          == regionsInfoReq.getRegionType()) {
+                        buildTRegionsInfo(
+                            tRegionInfosList, storageGroup, replicaSet, regionInfoMap);
+                      }
+                      if (regionsInfoReq.getRegionType() == -1) {
+                        buildTRegionsInfo(
+                            tRegionInfosList, storageGroup, replicaSet, regionInfoMap);
+                      }
+                    }));
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:717]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 10]
```diff
@@ -685,6 +691,62 @@ public void processLoadSnapshot(File snapshotDir) throws TException, IOException
...
+    for (TDataNodeLocation dataNodeLocation : replicaSet.getDataNodeLocations()) {
+      dataNodeIdList = Arrays.asList(dataNodeLocation.getDataNodeId());
+      rpcAddressList = Arrays.asList(dataNodeLocation.getExternalEndPoint().getIp());
+      rpcPortList = Arrays.asList(dataNodeLocation.getExternalEndPoint().getPort());
+    }
+    tRegionInfos.setDataNodeId(dataNodeIdList);
+    tRegionInfos.setRpcAddresss(rpcAddressList);
+    tRegionInfos.setRpcPort(rpcPortList);
+    // TODO: Wait for data migration. And then add the state
+    tRegionInfos.setStatus("Up");
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:746]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 11]
```diff
@@ -685,6 +691,62 @@ public void processLoadSnapshot(File snapshotDir) throws TException, IOException
     }
   }
 
+  public DataSet getRegionInfos(GetRegionsInfoReq regionsInfoReq) {
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java:694]: fixed~

File: server/src/test/resources/datanode4conf/iotdb-engine.properties [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+rpc_address=0.0.0.0
```
[the remark this answers is not in the document]
Comment (RYH61) [reply server/src/test/resources/datanode4conf/iotdb-engine.properties:20]: fixed~

File: server/src/test/resources/datanode4conf/logback.xml [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+<configuration scan="true" scanPeriod="60 seconds">
```
[the remark this answers is not in the document]
Comment (RYH61) [reply server/src/test/resources/datanode4conf/logback.xml:22]: fixed~

File: server/src/test/resources/datanode4conf/iotdb-metric.yml [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+# whether enable the module
```
[the remark this answers is not in the document]
Comment (RYH61) [reply server/src/test/resources/datanode4conf/iotdb-metric.yml:20]: My mistake, I have deleted it

File: thrift-confignode/src/main/thrift/confignode.thrift [hunk 15]
```diff
@@ -229,6 +229,27 @@ struct TDropFunctionReq {
   1: required string udfName
 }
 
+// show regions
+struct TShowRegionReq {
+  1: optional i32 regionType;
+}
+
+struct TRegionInfos {
```
[the remark this answers is not in the document]
Comment (RYH61) [reply thrift-confignode/src/main/thrift/confignode.thrift:237]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 16]
```diff
@@ -361,6 +366,48 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(TConsensu
     return result;
   }
 
+  public void getRegionInfos(
+      GetRegionLocationsReq regionsInfoReq, List<TRegionLocation> tRegionInfosList) {
+    regionInfoMap.forEach(
+        (consensusGroupId, regionInfoMap) -> {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:372]: `(consensusGroupId, regionGroup)`

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 17]
```diff
@@ -361,6 +366,48 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(TConsensu
...
+            buildTRegionsInfo(tRegionInfosList, storageGroupName, replicaSet, regionInfoMap);
+          }
+        });
+  }
+
+  private void buildTRegionsInfo(
+      List<TRegionLocation> tRegionInfosList,
+      String storageGroup,
+      TRegionReplicaSet replicaSet,
+      RegionGroup regionInfoMap) {
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java:388]: Rename _regionInfoMap_ to _regionGroup_

File: thrift-commons/src/main/thrift/common.thrift [hunk 18]
```diff
@@ -77,6 +77,17 @@ struct TDataNodeLocation {
...
 }
 
+struct TRegionLocation {
+  1: required i32 regionId;
+  2: required i64 regionType;
+  3: required string storageGroup
+  4: required i32 dataNodeId;
+  5: required string rpcAddresss;
+  6: required i32 rpcPort;
+  7: required i32 slots;
```
Comment (CRZbulabula) [inline thrift-commons/src/main/thrift/common.thrift:87]: The type of parameter _slots_ should be i64, sorry for didn't explain clearly last night.

File: thrift-confignode/src/main/thrift/confignode.thrift [hunk 19]
```diff
@@ -229,6 +229,16 @@ struct TDropFunctionReq {
   1: required string udfName
 }
 
+// show regions
+struct TShowRegionReq {
+  1: optional i32 regionType;
```
Comment (CRZbulabula) [inline thrift-confignode/src/main/thrift/confignode.thrift:234]: TConsensusGroupType

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java [hunk 20]
```diff
@@ -685,6 +691,62 @@ public void processLoadSnapshot(File snapshotDir) throws TException, IOException
...
+    tRegionInfos.setRegionId(replicaSet.getRegionId().getId());
+    tRegionInfos.setRegionType(replicaSet.getRegionId().getType().getValue());
+    tRegionInfos.setStorageGroup(storageGroup);
+    long slots = regionInfoMap.getCounter();
+    tRegionInfos.setSlots((int) slots);
+    for (TDataNodeLocation dataNodeLocation : replicaSet.getDataNodeLocations()) {
+      dataNodeIdList = Arrays.asList(dataNodeLocation.getDataNodeId());
+      rpcAddressList = Arrays.asList(dataNodeLocation.getExternalEndPoint().getIp());
+      rpcPortList = Arrays.asList(dataNodeLocation.getExternalEndPoint().getPort());
+    }
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/PartitionInfo.java]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java (hunk 16 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 21]
```diff
@@ -361,6 +366,48 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(TConsensu
...
+          }
+          if (regionsInfoReq.getRegionType() == TConsensusGroupType.PartitionRegion) {
+            buildTRegionsInfo(tRegionInfosList, storageGroupName, replicaSet, regionInfoMap);
+          }
+        });
+  }
+
+  private void buildTRegionsInfo(
+      List<TRegionLocation> tRegionInfosList,
+      String storageGroup,
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java (hunk 17 above)
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java]: fixed~

File: thrift-commons/src/main/thrift/common.thrift [hunk 22]
```diff
@@ -77,6 +77,17 @@ struct TDataNodeLocation {
   6: required TEndPoint schemaRegionConsensusEndPoint
 }
 
+struct TRegionLocation {
+  1: required i32 regionId;
+  2: required i64 regionType;
```
[the remark this answers is not in the document]
Comment (RYH61) [reply thrift-commons/src/main/thrift/common.thrift]: fixed~

File: thrift-commons/src/main/thrift/common.thrift (hunk 18 above)
Comment (RYH61) [reply thrift-commons/src/main/thrift/common.thrift]: fixed~

File: thrift-confignode/src/main/thrift/confignode.thrift (hunk 19 above)
Comment (RYH61) [reply thrift-confignode/src/main/thrift/confignode.thrift]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java [hunk 23]
```diff
@@ -361,6 +366,48 @@ public List<Pair<Long, TConsensusGroupId>> getSortedRegionSlotsCounter(TConsensu
...
+    tRegionInfos.setRegionId(replicaSet.getRegionId().getId());
+    tRegionInfos.setRegionType(replicaSet.getRegionId().getType().getValue());
+    tRegionInfos.setStorageGroup(storageGroup);
+    long slots = regionInfoMap.getCounter();
+    tRegionInfos.setSlots((int) slots);
+    for (TDataNodeLocation dataNodeLocation : replicaSet.getDataNodeLocations()) {
+      dataNodeIdList = Arrays.asList(dataNodeLocation.getDataNodeId());
+      rpcAddressList = Arrays.asList(dataNodeLocation.getExternalEndPoint().getIp());
+      rpcPortList = Arrays.asList(dataNodeLocation.getExternalEndPoint().getPort());
+    }
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/persistence/partition/StorageGroupPartitionTable.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/mpp/common/header/HeaderConstant.java [hunk 24]
```diff
@@ -200,4 +209,19 @@ public class HeaderConstant {
...
+
+  static {
+    showRegionHeader =
+        new DatasetHeader(
+            Arrays.asList(
+                new ColumnHeader(COLUMN_REGION_ID, TSDataType.INT32),
+                new ColumnHeader(COLUMN_TYPE, TSDataType.TEXT),
+                new ColumnHeader(COLUMN_STATUS, TSDataType.TEXT),
+                new ColumnHeader(COLUMN_STORAGE_GROUP, TSDataType.TEXT),
+                new ColumnHeader(COLUMN_SLOTS, TSDataType.INT32),
```
Comment (CRZbulabula) [inline server/src/main/java/org/apache/iotdb/db/mpp/common/header/HeaderConstant.java:221]: CLOUMN_SLOTS's datatype should be TSDataType.INT64?

**Review by CRZbulabula — COMMENTED** [review]
PTAL~

File: server/src/main/java/org/apache/iotdb/db/mpp/common/header/HeaderConstant.java (hunk 24 above)
Comment (RYH61) [reply server/src/main/java/org/apache/iotdb/db/mpp/common/header/HeaderConstant.java]: fixed~

**Review by CRZbulabula — APPROVED** [review]
LGTM!

**Review by neuyilan — APPROVED** [review]
LGTM

**Review by wangchao316 — APPROVED** [review]
LGTM

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java [hunk 25]
```diff
@@ -2217,4 +2219,19 @@ public Statement visitFlush(IoTDBSqlParser.FlushContext ctx) {
...
+
+  @Override
+  public Statement visitShowRegion(IoTDBSqlParser.ShowRegionContext ctx) {
+    ShowRegionStatement showRegionStatement = new ShowRegionStatement();
+    if (ctx.DATA() != null) {
+      showRegionStatement.setRegionType(TConsensusGroupType.DataRegion);
+    } else if (ctx.SCHEMA() != null) {
+      showRegionStatement.setRegionType(TConsensusGroupType.SchemaRegion);
+    } else {
+      showRegionStatement.setRegionType(TConsensusGroupType.PartitionRegion);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java:2233]: Set to Null? Maybe we will add a show partition region in the future.

**Review by qiaojialin — APPROVED** [review]
Could we add an IT for this? Maybe assert some columns that we are sure. Such as Region type or just no exception

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/config/executor/StandaloneConfigTaskExecutor.java [hunk 26]
```diff
@@ -256,4 +257,15 @@ public SettableFuture<ConfigTaskResult> showTTL(ShowTTLStatement showTTLStatemen
...
   }
+
+  @Override
+  public SettableFuture<ConfigTaskResult> showRegion(ShowRegionStatement showRegionStatement) {
+    SettableFuture<ConfigTaskResult> future = SettableFuture.create();
+    future.setException(
+        new StatementExecutionException(
+            RpcUtils.getStatus(
+                TSStatusCode.EXECUTE_STATEMENT_ERROR,
+                "Executing this command in standalone mode is not supported")));
```
[the remark this answers is not in the document]
Comment (RYH61) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/execution/config/executor/StandaloneConfigTaskExecutor.java]: fixed~

File: docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md [hunk 27]
```diff
@@ -111,3 +111,66 @@ KILL QUERY <queryId>
...
 
 其中 statement 最大显示长度为 64 字符。对于超过 64 字符的查询语句，将截取部分进行显示。
+
+## 集群 Region 分布监控工具
+
+集群中以 Region 作为数据复制和数据管理的单元，Region 的状态和分布对于系统运维和测试有很大帮助，如以下场景：
+
+- 查看集群中各个 Region 被分配到了哪些 DataNode，是否均衡
+
+当前 IoTDB 支持使用如下 SQL 查看Region：
```
[the remark this answers is not in the document]
Comment (RYH61) [reply docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md]: fixed~

File: docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md [hunk 28]
```diff
@@ -111,3 +111,66 @@ KILL QUERY <queryId>
...
+- `SHOW REGIONS`: 展示所有 Region
+- `SHOW SCHEMA REGIONS`: 展示所有 SchemaRegion 分布
+- `SHOW DATA REGIONS`: 展示所有 DataRegion 分布
+
+```sql
+IoTDB> show regions
++--------+------------+------+-------------+-----+----------+----------+----+
+|RegionId|        Type|Status|storage group|Slots|DataNodeId|HostAdress|Port|
++--------+------------+------+-------------+-----+----------+----------+----+
+|       5|  DataRegion|    Up|      root.sg|    1|         2| 127.0.0.1|6671|
```
[the remark this answers is not in the document]
Comment (RYH61) [reply docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md]: fixed~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 29]
```diff
@@ -644,6 +646,18 @@ public UDFManager getUDFManager() {
     return udfManager;
   }
 
+  @Override
+  public DataSet showRegion(GetRegionLocationsReq getRegionsinfoReq) {
+    TSStatus status = confirmLeader();
+    if (status.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+      return partitionManager.getRegionInfos(getRegionsinfoReq);
```
[the remark this answers is not in the document]
Comment (RYH61) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java]: fixed~

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java (hunk 25 above)
Comment (RYH61) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java]: fixed~

**Pull request merged**